src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs

```diff
@@ -74,119 +87,234 @@ public override IReadOnlyList<ProjectSnapshot> Projects
             }
         }
 
-        public DefaultProjectSnapshot FindProject(ProjectId id)
+        public override Workspace Workspace { get; }
+
+        public override void ProjectUpdated(ProjectSnapshotUpdateContext update)
         {
-            if (id == null)
+            if (update == null)
             {
-                throw new ArgumentNullException(nameof(id));
+                throw new ArgumentNullException(nameof(update));
             }
+            
+            if (_projects.TryGetValue(update.WorkspaceProject.FilePath, out var original))
+            {
+                if (!original.IsInitialized)
+                {
+                    // If the project has been uninitialized, just ignore the update. 
+                    return;
+                }
 
-            _projects.TryGetValue(id, out var project);
-            return project;
-        }
+                // This is an update to the project's computed values, so everything should be overwritten
+                var snapshot = original.WithComputedUpdate(update);
+                _projects[update.WorkspaceProject.FilePath] = snapshot;
 
-        public override Workspace Workspace { get; }
+                if (snapshot.IsDirty)
+                {
+                    // It's possible that the snapshot can still be dirty if we got a project update while computing state in
+                    // the background. We need to trigger the background work to asynchronously compute the effect of the updates.
+                    NotifyBackgroundWorker(snapshot.CreateUpdateContext());
+                }
+
+                // Now we need to know if the changes that we applied are significant. If that's the case then 
+                // we need to notify listeners.
+                if (snapshot.HasChangesComparedTo(original))
+                {
+                    NotifyListeners(new ProjectChangeEventArgs(snapshot, ProjectChangeKind.Changed));
+                }
+            }
+        }
 
-        public override void ProjectAdded(Project underlyingProject)
+        public override void HostProjectAdded(HostProject hostProject)
         {
-            if (underlyingProject == null)
+            if (hostProject == null)
             {
-                throw new ArgumentNullException(nameof(underlyingProject));
+                throw new ArgumentNullException(nameof(hostProject));
             }
 
-            var snapshot = new DefaultProjectSnapshot(underlyingProject);
-            _projects[underlyingProject.Id] = snapshot;
+            // It's possible that the workspace project was already initialized, and the host project is showing up
```

Comment:
Good to be extra cautious but out of curiosity when would this be possible? From my understanding Roslyn reacts to the pieces being added to the host project system.

Reply:
Another design idea that I had was to make it so that you **always** have a HostProject when you have a snapshot. That simplifies things quite a bit.
 
Basically you may or may not have a Workspace Project, but you always have a HostProject. Thoughts?

Reply:
> Basically you may or may not have a Workspace Project, but you always have a HostProject. Thoughts?

Yup, that's what I was getting at 😄

Reply:
@NTaylorMullen - so let's make a decision then.. Should we discuss this tomorrow when I get into the office or are you convinced?

Mainly I want to have confidence that building the VS4Mac version of this is not going to change the design **again**. Can you take a gander at what you have to work with in VS4Mac and make a judegement call?

Reply:
> Should we discuss this tomorrow when I get into the office or are you convinced?

I'm convinced. I'm just curious how you want to do it is all. All I know is that I don't want to bring back `CreateDesignTime` on `RazorEngine` / `RazorProjectEngine` 😄